File: 05/CSharp/ExpressionApp/BinaryOperator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ExpressionApp
{
    abstract class BinaryOperator : IExpression
    {
        protected IExpression fOp;
        protected IExpression sOp;

        abstract public double Evaluate(double param);
    }
}

File: 07/Java/TripleExpression/BinaryOperator.java
package TripleExpression;

abstract public class BinaryOperator implements TripleExpression {
    protected TripleExpression fOp;
    protected TripleExpression sOp;
    
    BinaryOperator(TripleExpression first, TripleExpression second) {
        fOp = first;
        sOp = second;
    }

    abstract protected int apply(int a, int b) throws Exception;
    
    public int evaluate(int x, int y, int z) throws Exception {
        return apply(fOp.evaluate(x, y, z), sOp.evaluate(x, y, z));
    }
}